worlds/cv64/__init__.py

```diff
@@ -0,0 +1,310 @@
+from .rom import LocalRom, patch_rom, get_base_rom_path, CV64DeltaPatch
+from .client import Castlevania64Client
+
+
+class CV64Settings(settings.Group):
+    class RomFile(settings.UserFilePath):
+        """File name of the CV64 US 1.0 rom"""
+        copy_to = "Castlevania (USA).z64"
+        description = "CV64 (US 1.0) ROM File"
+        md5s = [CV64DeltaPatch.hash]
+
+    rom_file: RomFile = RomFile(RomFile.copy_to)
+
+
+class CV64Web(WebWorld):
+    theme = "stone"
+
+    tutorials = [Tutorial(
+        "Multiworld Setup Guide",
+        "A guide to setting up the Archipleago Castlevania 64 randomizer on your computer and connecting it to a "
+        "multiworld.",
+        "English",
+        "setup_en.md",
+        "setup/en",
+        ["Liquid Cat"]
+    )]
+
+
+class CV64World(World):
+    """
+    Castlevania for the Nintendo 64 is the first 3D game in the franchise. As either whip-wielding Belmont descendant
+    Reinhardt Schneider or powerful sorceress Carrie Fernandez, brave many terrifying traps and foes as you make your
+    way to Dracula's chamber and stop his rule of terror!
+    """
+    game = "Castlevania 64"
+    item_name_groups = {
+        "Bomb": {iname.magical_nitro, iname.mandragora},
+        "Ingredient": {iname.magical_nitro, iname.mandragora},
+    }
+    location_name_groups = {stage: set(get_locations_from_stage(stage)) for stage in vanilla_stage_order}
+    options_dataclass = CV64Options
+    options: CV64Options
+    settings: typing.ClassVar[CV64Settings]
+    topology_present = True
+    data_version = 1
+
+    item_name_to_id = get_item_names_to_ids()
+    location_name_to_id = get_location_names_to_ids()
+
+    active_stage_exits: typing.Dict[str, typing.Dict]
+    active_stage_list: typing.List[str]
+    active_warp_list: typing.List[str]
+    reinhardt_stages: bool = True
+    carrie_stages: bool = True
+    branching_stages: bool = False
+    starting_stage: str = rname.forest_of_silence
+
+    total_s1s: int = 7
+    s1s_per_warp: int = 1
+    total_s2s: int = 0
+    required_s2s: int = 0
+    drac_condition: int = 0
+
+    auth: bytearray
+
+    web = CV64Web()
+
+    @classmethod
+    def stage_assert_generate(cls, multiworld: MultiWorld) -> None:
+        rom_file = get_base_rom_path()
+        if not os.path.exists(rom_file):
+            raise FileNotFoundError(rom_file)
+
+    def generate_early(self) -> None:
+        # Generate the player's unique authentication
+        self.auth = bytearray(self.multiworld.random.getrandbits(8) for _ in range(16))
+
+        self.total_s1s = self.options.total_special1s.value
+        self.s1s_per_warp = self.options.special1s_per_warp.value
+        self.drac_condition = self.options.draculas_condition.value
+
+        # If there are more S1s needed to unlock the whole warp menu than there are S1s in total, drop S1s per warp to
+        # something manageable.
+        while self.s1s_per_warp * 7 > self.total_s1s:
+            self.s1s_per_warp -= 1
```

Comment:
I think it would make sense to print something or indicate in some way that you are, effectively, changing the settings somebody used.

Reply:
There's definitely something I don't understand going on here, because what happens if `s1s_per_warp` is 1 and `self.total_s1s` is also 1? Wouldn't this make `s1s_per_warp` 0? Which isn't something allowed in the yaml at least. I feel like I'm missing something about what is going on here

Reply:
I'm also note sure why this is using a `while` instead of just `self.s1s_per_warp = self.total_s1s // 7`

Reply:
Invalid settings now warn the player and adjust the setting value, so it shows in the spoiler.

Reply:
I think my change to an `if` block instead of a `while` is still a good idea

Reply:
s1s_per_warp at 0 works surprisingly better than I thought it would! XD
Maybe down the line I will turn it into an actual setting, but right now, no, this is not intended.

![Screenshot 2024-03-10 193508](https://github.com/ArchipelagoMW/Archipelago/assets/74896918/30bcb6ed-d3fa-4ecd-99ce-a66a1cbca4a5)